Add the navigation API

domenic opened the pull request:
This imports much of the specification from https://github.com/WICG/navigation-api/blob/91c2e7f959418d71e26e5b9a6723fed2944001d9/spec.bs. It includes some fixes and renamings, as well as a lot of rearranging to flow better, but the feature set is the same.

In addition to the new "navigation API" section under "APIs related to navigation and session history", this makes the following changes to integrate the navigation API:

* Introduces the NavigationHistoryBehavior enum, as a superset of the "history handling" specification type. This includes an "auto" value to reflect that existing entry points will sometimes default to "replace" instead of "push". The navigation API allows overriding that auto behavior in some circumstances by explicitly supplying "push". (But in some cases, i.e., the initial `about:blank` or `javascript:` URLs, this is still not allowed.)

* Introduces the concept of "user navigation involvement". This will be helpful for solving issues such as https://github.com/w3c/webappsec-fetch-metadata/issues/71.

* Introduces the concept of "history-action user activation", which is a separate type of user activation meant to be used specifically to prevent back button tracking. It has no associated timeout (unlike transient activation), but can be flipped to false (unlike sticky activation). It is used by the navigation API to prevent calling `navigateEvent.preventDefault()` on two traversals in a row without intervening user activation. It is likely also to be useful for issues such as https://github.com/whatwg/html/issues/7832.

* The activation behavior for `<a>` and `<area>` elements is made more rigorous and consolidated into one place, so that we can hook into it appropriately to fire `navigate` events.

* Some surgery was needed on "apply the history step". The result is that it now has slightly more parameters, but also several wrapper algorithms which take care of setting up those parameters correctly for the cases of: navigable creation/destruction, push/replace, reload, and traverse. It also returns a value indicating if the application of the history step was canceled, and if so, how; this is used to give informative errors as return values of `navigation.traverseTo()`.

* The "check if unloading is user-canceled" algorithm has become "check if unloading is canceled", as it now handles firing a possibly-cancelable `navigate` event at the top-level traversable.

Other changes scattered throughout are mostly integrating appropriate calls to the navigation API as necessary, especially to fire `navigate` events.

---

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * [Mozilla](https://mozilla.github.io/standards-positions/#navigation-api)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/navigation-api?label=experimental&label=master&aligned (to be moved into html/browsers/ after this PR is merged)
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: https://bugs.chromium.org/p/chromium/issues/detail?id=1183545
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1777171
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=258384
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): N/A
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): N/A
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/pull/21960


(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/8502/acknowledgements.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/acknowledgements.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/acknowledgements.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/browsing-the-web.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/browsing-the-web.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/browsing-the-web.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/document-lifecycle.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/document-lifecycle.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/document-lifecycle.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/document-sequences.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/document-sequences.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/document-sequences.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/form-control-infrastructure.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/form-control-infrastructure.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/form-control-infrastructure.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/form-elements.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/form-elements.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/forms.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/forms.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/forms.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/iframe-embed-object.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/iframe-embed-object.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/iframe-embed-object.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/image-maps.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/image-maps.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/image-maps.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/index.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/index.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/index.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/indices.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/indices.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/indices.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/infrastructure.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/infrastructure.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/input.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/input.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/input.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/interaction.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/interaction.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/interaction.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/links.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/links.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/links.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/nav-history-apis.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/nav-history-apis.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/nav-history-apis.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/parsing.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/parsing.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/parsing.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/popover.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/popover.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/popover.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/semantics.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/semantics.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/semantics.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/server-sent-events.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/server-sent-events.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/server-sent-events.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/text-level-semantics.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/text-level-semantics.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/text-level-semantics.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )
<a href="https://whatpr.org/html/8502/webappapis.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">/webappapis.html</a>  ( <a href="https://whatpr.org/html/8502/8a32a33...40c167d/webappapis.html" title="Last updated on Jun 22, 2023, 3:00 AM UTC (40c167d)">diff</a> )

annevk reviewed (commented):
Thanks for your work on this! Editorially it looks great.

I did an initial pass. I think this definitely needs review from multiple people given the size. I put my thoughts here as doing it line-by-line would probably be unwieldy.

* Can download the hyperlink still be reconciled with the navigate algorithm? (Looks like yes.)
* Nit: i.e. and e.g. trailing commas.
* Why "classic" for "classic history API state" and not "legacy"?
* Nit: I find the three letter variable names rather confusing. I would prefer sessionHistoryEntry or entry over she, for instance.
* "navigation API method tracker-derived result" doesn't initialize _apiMethodTracker's_.
* Missing word: "that are not via navigation API methods"
* NavigationDestination isn't linked from IDL
* The NavigateEvent's type attribute is never initialized. (This is a bit concerning as I guess it means the implementation wasn't written based on the specification.)
* "inner navigate event firing algorithm" has a lot of magic around influencing the normal order of the event loop. Is this really unique to navigation or is there a deeper issue here?
* "user navigation involvement" seems problematic in that isTrusted can be true for events dispatched by the user agent. Consider `onmessage = e => console.log(e.isTrusted);postMessage("x")`. `click()` setting it to false seems more of an exception. Did not look at what `submit()` does here. It might be okay for a limited set of call sites, but then we ought to make that clear somehow. Perhaps by inlining it.
* This doesn't impact BFCache, right?

domenic commented:
Thanks so much for taking a look!!

> I did an initial pass. I think this definitely needs review from multiple people given the size.

I've enrolled @jakearchibald to help :)

> Can download the hyperlink still be reconciled with the navigate algorithm? (Looks like yes.)

Yep, this doesn't significantly change things, and ensures anything added to navigation is also added to downloading.

> Why "classic" for "classic history API state" and not "legacy"?

Similarly with classic scripts, it felt too early to call the thing people are using a lot in today's code "legacy". I don't mind changing it though, if you'd prefer :).

> The NavigateEvent's type attribute is never initialized. (This is a bit concerning as I guess it means the implementation wasn't written based on the specification.)

It is; see "inner navigate event firing algorithm" step 9.

> "inner navigate event firing algorithm" has a lot of magic around influencing the normal order of the event loop. Is this really unique to navigation or is there a deeper issue here?

I think it's not unique to navigation, but might be exotic enough to have never occured before. The reason this comes up is that we have an algorithm that:

- Sometimes is called with an empty JS stack (in our case, traversal) and sometimes with a full JS stack (non-traverse navigation);
- Both fires events and resolves promises;
- And we generally expect developers to use both those events and promises, and expect a consistent ordering between them.

This _sort of_ comes up with event handlers and users clicking (empty stack) vs. `click()` (no clicking), as discussed in https://jakearchibald.com/2015/tasks-microtasks-queues-and-schedules/. But there's no platform-generated promises in that case. If we added some sort of `document.nextClick` promise, then maybe we'd be in a similar situation. And in that case we might want to think about abstracting things. But for now I think it's OK to do this as a one-off.

> "user navigation involvement" seems problematic in that isTrusted can be true for events dispatched by the user agent. [..] It might be okay for a limited set of call sites, but then we ought to make that clear somehow. Perhaps by inlining it.

Fair point; I'll have a think about the best way to do this.

> This doesn't impact BFCache, right?

It doesn't prevent bfcaching, or have any ongoing processes that need to be shut down or managed during bfcaching. But it does have one interesting interaction, which is cases like this:

1. Session history is A -> B -> C -> D, you are currently on B.
2. You move to C, and keep B around in bfcache.
3. You do `history.pushState()` or `navigation.navigate()` or anything else, blowing away D and replacing it with E.
4. You do `history.back()` or `navigation.back()`, sending you back to B, reactivating it from bfcache.

We need to ensure that `navigation.entries()` contains A -> B -> C -> E, not its old value of A -> B -> C -> D. Also we need to ensure a `dispose` event is fired for D. This is accomplished in "update the entries for reactivation".

domenic commented on source:
```diff
@@ -93826,19 +97585,80 @@ location.href = '#foo';</code></pre>
+
+   <li><p>Return the result of <span data-x="apply the history step">applying the history
+   step</span> <var>step</var> to <var>traversable</var> given false, false, null, null, and
+   null.</p></li>
```
I think what I've done here, of putting the definition of "apply the history step" right below all its callers (which are now all wrapper algorithms), is a reasonable way of dealing with this.

domenic commented on source:
```diff
@@ -88583,8 +88699,3216 @@ interface <dfn interface>History</dfn> {
+   </li>
+  </ol>
+
+  <p>To <dfn>update the entries for a same-document navigation</dfn> given a <code>Navigation</code>
```
I'll add a pointer to https://github.com/whatwg/html/issues/8620

domenic commented:
> then the timing of the "dispose" event is linked to the other document, and that timing will be incorrect when observed by the parent window.

What do you mean by incorrect?

Perhaps relevant: since all such documents are in the same agent, they share the same event loop. And all tasks use the same "navigation and traversal task source". I'm not sure that's enough to tackle what you're thinking of, but worth keeping in mind...

domenic commented:
Thanks Jake, your understanding is correct. We discussed this behavior in some detail in https://github.com/WICG/navigation-api/issues/95#issuecomment-902314407 and some meetings surrounding it. (See also https://github.com/WICG/navigation-api/pull/157#issuecomment-912754893.) I don't think there's a great option here for satisfying both the within-document cases, and the grabbing-someone-else's-`entries()` cases. Since the latter is kind of against the spirit of the navigation API, I think it's OK for the behavior to be suboptimal in that case.

Maybe we should add some discussion about this, either in the spec or MDN, to make the behavior of `dispose` + bfcache a bit clearer.

jakearchibald commented:
If it's intentional, that's fine. I guess the developer guidance would be "the dispose event is only reliable when observed from that navigable, not child, parent navigables, or other traversables"

tbondwilkinson commented:
Just want to leave a comment that the timing of `dispose` isn't important for most use cases I believe, and I think if it has the potential to cause confusion, `dispose` could be limited in scope to just same-document.

tbondwilkinson commented on source:
```diff
@@ -88583,8 +88705,3221 @@ interface <dfn interface>History</dfn> {
+    settings object</span>.</p>
+
+    <p class="note">Per the <a href="#note-suppress-microtasks-during-navigation-events">previous
+    note</a>, this stops suppress any potential promise handler microtasks, causing them to run at
```
this stops "suppressing"

I think

tbondwilkinson commented on source:
```diff
@@ -88583,8 +88705,3221 @@ interface <dfn interface>History</dfn> {
+   <dt><code data-x=""><var>event</var>.<span subdfn data-x="dom-NavigationCurrentEntryChangeEvent-navigationType">navigationType</span></code></dt>
+   <dd><p>Returns the type of navigation which caused the current entry to change, or null if the
+   change is due to <code
+   data-x="dom-Navigation-updateCurrentEntry">navigation.updateCurrentEntry()</code>.</p></dd>
```
It's a bit odd that the navigationType is null for updateState(). Are we sure we don't want to throw in an extra navigation type for that?

domenic replied:
I think since it's very much not a navigation, it's best to have it be null, to signal the exceptional nature of what's going on.

tbondwilkinson reviewed (commented):
Just nits really. This is a massive rewrite, really impressive seeing it all laid out!

domenic commented on source:
```diff
@@ -88583,8 +88705,3221 @@ interface <dfn interface>History</dfn> {
   </div>
 
 
+  <h4 id="navigation-api">The navigation API</h4>
```
This section contains more than just the `Navigation` interface, as it's the section header for all of the navigation API. IN that way it's more like the section titles elsewhere of "Dynamic markup insertion", "User prompts", "Server-sent events", "Cross-document messaging", etc. The addition of "API" is a bit unusual, but it's necessary because of the generic term we've chosen: throughout the spec we need to distinguish "navigation" and "the navigation API", whereas we don't need to distinguish "WebSockets" and "the WebSockets API".

domenic commented on source:
```diff
@@ -88583,8 +88705,3221 @@ interface <dfn interface>History</dfn> {
+     immediately.</p></li>
+
+     <li><p>For cross-document navigations, or navigations that result in 204 or 205 <span
+     data-x="concept-response-status">statuses</span> or `<code
```
Yep, this is https://infra.spec.whatwg.org/#byte-sequences

domenic commented on source:
```diff
@@ -88583,8 +88705,3221 @@ interface <dfn interface>History</dfn> {
+    <p>If <span>this</span>'s <span>upcoming non-traverse API method tracker</span> is
+    <var>apiMethodTracker</var>, then:</p>
+
+    <p class="note">This means the <span>navigate</span> algorithm bailed out before ever getting to
```
Thanks; I tweaked the note similar to what you suggested.

As for why not modifying the navigate algorithm, it seems cleaner to keep the navigation API-specific stuff a bit separated, especially since it's important only for this one call site. Navigate is called by many different parts of the spec, and having a return value that 90% of call sites ignore seems a bit unfortunate.

smaug---- commented on source:
```diff
@@ -1794,11 +1794,15 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    document">focused area</span> to <var>document</var>'s <span>viewport</span>, and set
+    <var>document</var>'s <span>relevant global object</span>'s <span
+    data-x="window-navigation-api">navigation API</span>'s <span>focus changed during ongoing
+    navigation</span> to false.</p>
```
This is surprising. Why would removing a node clear such flag.

domenic replied:
So how `focusReset: "after-transition"` is supposed to work is as follows:

1. If nothing relevant changes in the document, focus gets reset after the transition
2. If the user (or developer, but the user is more important) changes focus during the transition, we _don't_ reset the focus: that would be too disruptive.
3. But, if the user or developer changes focus during the transition, and then the developer removes the currently-focused DOM node from the document, we _do_ reset the focus.

Without this step you're commenting on, (3) will be missing.

Note here "reset the focus" means the slightly-special behavior in https://whatpr.org/html/8502/nav-history-apis.html#potentially-reset-the-focus , of using the autofocus behavior and also resetting the sequential focus navigation starting point. So it's more than just the focus fixup steps.

What do you think?

smaug---- commented on source:
```diff
@@ -88789,8 +88911,3228 @@ interface <dfn interface>History</dfn> {
+   return true.</p></li>
+
+   <li><p>If <var>document</var>'s <span data-x="concept-document-origin">origin</span> is <span
+   data-x="concept-origin-opaque">opaque</span>, then return true.</p></li>
```
This looks surprising. One couldn't use the API for top level opaque documents for example.

domenic replied:
Yes. We haven't run into people using top-level opaque documents who complained, but it's something we could loosen up in the future if you are in contact with some developers who would like to use this API, and can't.

smaug---- commented on source:
```diff
@@ -88789,8 +88911,3228 @@ interface <dfn interface>History</dfn> {
+   <li><p>If <var>navigation</var>'s <span>upcoming traverse API method
+   trackers</span>[<var>key</var>] <span data-x="map exists">exists</span>, then return a
+   <span>navigation API method tracker-derived result</span> for <var>navigation</var>'s
+   <span>upcoming traverse API method trackers</span>[<var>key</var>].</p></li>
```
Isn't this surpising? if the key is already there, reuse that even if  the map might have gotten other entries after the key was added.

domenic replied:
I don't understand your second sentence, or what's surprising about this behavior. Could you say more?

To be clear, this is about cases like:

```js
const key = navigation.entries()[navigation.currentEntry.index - 1];

const result1 = navigation.traverseTo(key);

// ... probably in some other part of the app ...
const result2 = navigation.traverseTo(key);
```

and says that `result2` should just reflect what's going on with `result1` (i.e. have the same promises). I'm not sure what other behavior you'd envision here...

smaug---- commented on source:
```diff
@@ -88789,8 +88911,3228 @@ interface <dfn interface>History</dfn> {
+  readonly attribute USVString <span data-x="dom-NavigationDestination-url">url</span>;
+  readonly attribute DOMString? <span data-x="dom-NavigationDestination-key">key</span>;
+  readonly attribute DOMString? <span data-x="dom-NavigationDestination-id">id</span>;
+  readonly attribute long long <span data-x="dom-NavigationDestination-index">index</span>;
```
What is the use case for the index? Why can't we have a pointer to the entry?
I think I'd rather not expose index attribute in any interface, just because they may be wrong after any async step. (that is also based on the implementation experience, dealing with indexes is very error prone.)

domenic replied:
Yep. If you're worried about them getting out of sync, you can implement this exactly like the spec does, and just do the O(n) `indexOf()` call on each getter invocation. Since n is small (<50, due to history pruning, and usually a good bit less, since it's same-origin contiguous) this shouldn't be a big deal.

smaug---- replied:
I am worried about webcompat. Different browsers have different max length, so index becomes invalid at different times in different implementations. I think index should not be exposed in the API.

domenic replied:
Index never becomes invalid. The index is well-defined in this specification and is the same as `indexOf()`; it is just sugar. There is no compat problem here that is any different from the existence of the `entries()` array itself.

smaug---- replied:
How does it not get out-of-sync? If one first  accesses the entry using the index and right after that does history.pushState() 50 or 100 times, and then uses index again, isn't the index pointing to wrong place?

domenic replied:
No. `index` becomes -1 at that point, per the definition in this spec.

smaug---- replied:
hmm, github keeps moving this comment for me, so I'm not sure which index attribute I initially added this for, but NavigationDestination.index doesn't change. NavigationDestination[index] is set just once. It is not dynamic.

domenic replied:
Oh, I see, sorry for the confusion! Yes, that's a great catch on an inconsistency. I've fixed that, and we'll work on adding tests and updating the Chromium implementation.

domenic commented on source:
```diff
@@ -76884,26 +76978,28 @@ END:VCARD</pre>
+   <dt><dfn export data-x="history-action activation-consuming API" data-lt="history-action activation-consuming API">History-action activation-consuming APIs</dfn></dt>
+   <dd><p>These APIs require the <span>history-action activation</span> state to be true, and they
+   <span>consume history-action user activation</span> in each call to prevent multiple calls per
+   user activation.</p></dd>
```
This is added to prevent back-trapping. See "traversals have special restrictions on canceling the navigation" in https://github.com/wicg/navigation-api#restrictions-on-firing-canceling-and-responding . It is also envisioned to be used by https://github.com/WICG/close-watcher .

zcorpan commented:
> @smaug---- from Gecko and @annevk (at Gecko at the time; now at Apple) were involved in the design of this API, and we have some positive sentiments in https://github.com/mozilla/standards-positions/issues/543. I'd love to have them confirm their support.

Mozilla is [positive](https://mozilla.github.io/standards-positions/#navigation-api).

domenic commented:
Thanks so much Simon!

Given that we've gotten review on this from multiple sources already, and all the checkboxes in the OP are filled out, I'll plan to merge this Monday (Japan time). Of course, if more reviews come in, I'm happy to delay.
